diff --git a/CaptureOverlay.qml b/CaptureOverlay.qml
index b9f3165..519ff68 100644
--- a/CaptureOverlay.qml
+++ b/CaptureOverlay.qml
@@ -25,6 +25,8 @@ SettingsOverlay {
     property int _recordingDuration: clock.enabled ? ((clock.time - _startTime) / 1000) : 0
     property var _startTime: new Date()
 
+    property alias zoomIndicator: zoomIndicator
+
     width: captureView.width
     height: captureView.height
 
diff --git a/CaptureView.qml b/CaptureView.qml
index 9555b47..565c5ca 100644
--- a/CaptureView.qml
+++ b/CaptureView.qml
@@ -8,6 +8,7 @@ import org.nemomobile.ngf 1.0
 import org.nemomobile.dbus 2.0
 import org.nemomobile.notifications 1.0
 import org.nemomobile.systemsettings 1.0
+import Nemo.Configuration 1.0
 
 import "../settings"
 
@@ -72,9 +73,8 @@ FocusScope {
         }
     }
 
-    property bool handleVolumeKeys: camera.imageCapture.ready
-                                    && keysResource.acquired
-                                    && camera.captureMode == Camera.CaptureStillImage
+    property bool handleVolumeKeys: keysResource.acquired
+                                    && ((camera.captureMode == Camera.CaptureStillImage) || (camera.captureMode == Camera.CaptureVideo))
                                     && !captureView._captureOnFocus
     property bool captureOnVolumeRelease
 
@@ -694,15 +694,18 @@ FocusScope {
     }
 
     Keys.onVolumeDownPressed: {
-        if (handleVolumeKeys && !event.isAutoRepeat) {
-            camera.lockAutoFocus()
-            captureOnVolumeRelease = true
+        if (handleVolumeKeys) {
+            camera.digitalZoom = Math.max(1, camera.digitalZoom - 1)
+            captureOverlay.zoomIndicator.show()
         }
     }
     Keys.onVolumeUpPressed: {
-        if (handleVolumeKeys && !event.isAutoRepeat) {
-            camera.lockAutoFocus()
-            captureOnVolumeRelease = true
+        if (handleVolumeKeys) {
+            camera.digitalZoom = Math.min(camera.maximumDigitalZoom, camera.digitalZoom + 1)
+            if (!event.isAutoRepeat) {
+                captureOverlay.zoomIndicator.maximumZoom = camera.maximumDigitalZoom
+            }
+            captureOverlay.zoomIndicator.show()
         }
     }
 
@@ -722,10 +725,13 @@ FocusScope {
             return
         }
 
-        if (event.key == Qt.Key_CameraFocus) {
-            camera.lockAutoFocus()
-        } else if (event.key == Qt.Key_Camera) {
-            captureView._triggerCapture() // key having half-pressed state too so can capture already here
+        if (event.key == Qt.Key_Camera) {
+            if (camera.captureMode == Camera.CaptureVideo) {
+                captureView._triggerCapture()
+            } else {
+                camera.lockAutoFocus()
+                captureOnVolumeRelease = true
+            }
         }
     }
 
@@ -738,18 +744,14 @@ FocusScope {
             return
         }
 
-        if (event.key == Qt.Key_CameraFocus) {
-            // note: forces capture if it was still pending. debatable if that should be allowed to finish.
-            camera.unlockAutoFocus()
-        } else if ((event.key == Qt.Key_VolumeDown || event.key == Qt.Key_VolumeUp)
-                   && captureOnVolumeRelease && handleVolumeKeys) {
+        if ((event.key == Qt.Key_Camera) && captureOnVolumeRelease && handleVolumeKeys) {
             captureView._triggerCapture()
         }
     }
 
     Permissions {
         enabled: captureView.activeFocus
-                    && camera.captureMode == Camera.CaptureStillImage
+                    && ((camera.captureMode == Camera.CaptureStillImage) || (camera.captureMode == Camera.CaptureVideo))
                     && camera.cameraState == Camera.ActiveState
         autoRelease: true
         applicationClass: "camera"
@@ -817,4 +819,23 @@ FocusScope {
     AboutSettings {
         id: aboutSettings
     }
+    
+    property int initialMediaKeysEnabled: 1
+
+    onVisibleChanged: {
+        if (!initialMediaKeysEnabled)
+            return
+
+        mediaKeysEnabled.value = !visible
+    }
+
+    ConfigurationValue {
+        id: mediaKeysEnabled
+        key: "/apps/photonq-helper/mediaKeysEnabled"
+        defaultValue: 1
+
+        Component.onCompleted: {
+            initialMediaKeysEnabled = mediaKeysEnabled.value
+        }
+    }
 }
